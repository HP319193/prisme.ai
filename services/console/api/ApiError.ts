import { APIError } from "../../../packages/types";

export class ApiError extends Error implements APIError {
  public code: number;
  public error: APIError["error"];
  public details: APIError["details"];

  constructor({ error, message, details }: APIError, code: number) {
    super(message);
    this.code = code;
    this.error = error;
    this.details = details;
  }

  toString() {
    return this.error;
  }
}

export default ApiError;
