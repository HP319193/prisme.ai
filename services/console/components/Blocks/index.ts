export { default as Title } from './Title';
export { default as Cards } from './Cards';
export { default as DataTable } from './DataTable';
