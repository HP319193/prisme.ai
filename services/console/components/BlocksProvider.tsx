import { FC, HTMLAttributes } from 'react';
import { Loading as DSLoading, SchemaForm } from '@prisme.ai/design-system';
import NextLink from 'next/link';
import externals from '../utils/externals';
import Image from 'next/image';
import { BlocksProvider as PRBlocksProvider } from '@prisme.ai/blocks';
import down from '../icons/down.svg';
import { usePageEndpoint } from '../utils/urls';

const Loading = () => (
  <DSLoading className="bg-white absolute top-0 right-0 bottom-0 left-0" />
);
const Link = ({
  href = '',
  ...props
}: { href: string } & HTMLAttributes<HTMLAnchorElement>) => {
  return <NextLink {...props} href={href || ''} />;
};
const DownIcon = ({ className }: { className?: string }) => (
  <Image src={down.src} width={14} height={14} alt="" className={className} />
);

export const BlocksProvider: FC = ({ children }) => {
  const pageHost = usePageEndpoint();
  return (
    <PRBlocksProvider
      externals={externals}
      components={{ Link, Loading, DownIcon, SchemaForm }}
      utils={{
        getWorkspaceHost() {
          return pageHost;
        },
      }}
    >
      {children}
    </PRBlocksProvider>
  );
};

export const PublicBlocksProvider: FC = ({ children }) => {
  return (
    <PRBlocksProvider
      externals={externals}
      components={{ Link, Loading, DownIcon, SchemaForm }}
    >
      {children}
    </PRBlocksProvider>
  );
};
