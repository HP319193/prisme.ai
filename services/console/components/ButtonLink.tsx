import Link from "next/link"
import { ButtonHTMLAttributes, FC } from "react"

interface ButtonLinkProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  href?: string;
}
export const ButtonLink: FC<ButtonLinkProps> = ({ children, href, ...props }) => {

  return <button style={{ background: 'none', border: 0, cursor: 'pointer', fontSize: 'inherit', ...props.style }} {...props}>
    {href && <Link href={href}><>{children}</></Link>}
    {!href && children}
  </button>
}

export default ButtonLink
