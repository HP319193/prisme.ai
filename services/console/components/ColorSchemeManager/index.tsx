import Storage from '../../utils/Storage';

export function getColorScheme() {
  return (
    (document.documentElement.dataset.colorScheme as 'dark' | 'light') ||
    Storage.get('color-scheme') ||
    (window.matchMedia('(prefers-color-scheme: dark)').matches
      ? 'dark'
      : 'light')
  );
}

export function setColorScheme(scheme: 'dark' | 'light', save: boolean = true) {
  if (save) {
    Storage.set('color-scheme', scheme);
  }
  document.documentElement.setAttribute('data-color-scheme', scheme);
}

export function toggleColorScheme() {
  const next = getColorScheme() === 'dark' ? 'light' : 'dark';
  setColorScheme(next);
}

export const ColorSchemeManager = () => {
  if (typeof window === 'undefined') return null;

  setColorScheme(getColorScheme(), false);

  // @ts-ignore
  window.Prisme = window.Prisme || {};
  // @ts-ignore
  window.Prisme.ai = window.Prisme.ai || {};
  // @ts-ignore
  window.Prisme.ai.colorScheme = {
    get: getColorScheme,
    set: setColorScheme,
    toggle: toggleColorScheme,
  };
  return null;
};

export default ColorSchemeManager;
