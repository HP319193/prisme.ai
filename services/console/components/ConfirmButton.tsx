import { Button, Tooltip } from '@prisme.ai/design-system';
import { ButtonProps } from '@prisme.ai/design-system/lib/Components/Button';
import { TooltipProps } from 'antd';
import { useTranslation } from 'next-i18next';
import { ReactNode, useRef } from 'react';

interface ConfirmButtonProps extends ButtonProps {
  children: ReactNode;
  onConfirm: () => void;
  confirmLabel?: string;
  yesLabel?: string;
  noLabel?: string;
}
export const ConfirmButton = ({
  children,
  onConfirm,
  confirmLabel,
  yesLabel,
  noLabel,
  ...props
}: ConfirmButtonProps) => {
  const tooltipRef = useRef<{ close: () => void }>(null);
  const { t } = useTranslation('common');
  return (
    <Tooltip
      ref={tooltipRef}
      title={
        <div>
          <div>{confirmLabel}</div>
          <div className="flex justify-end">
            <Button
              type="button"
              variant="grey"
              onClick={() => {
                onConfirm();
                tooltipRef.current?.close();
              }}
            >
              {yesLabel || t('yes')}
            </Button>
            <Button type="button" onClick={() => tooltipRef.current?.close()}>
              {noLabel || t('no')}
            </Button>
          </div>
        </div>
      }
      trigger={['click']}
      color="var(--warning-color)"
    >
      <Button className="!text-warning" {...props}>
        {children}
      </Button>
    </Tooltip>
  );
};
export default ConfirmButton;
