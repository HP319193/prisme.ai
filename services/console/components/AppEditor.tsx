import {
  Button,
  Loading,
  notification,
  Schema,
  SchemaForm,
} from '@prisme.ai/design-system';
import { BlockLoader } from '@prisme.ai/blocks';
import api from '../utils/api';
import useAppConfig from '../utils/useAppConfig';
import { useWorkspace } from './WorkspaceProvider';
import { useWorkspaceLayout } from '../layouts/WorkspaceLayout/context';
import { useTranslation } from 'next-i18next';
import { useCallback } from 'react';

interface AppEditorProps {
  schema?: Schema;
  block?: string;
  appId: string;
}

const AppEditor = ({ schema, block, appId }: AppEditorProps) => {
  const {
    workspace: { id: workspaceId },
  } = useWorkspace();
  const { t } = useTranslation('workspaces');
  const { t: commmonT } = useTranslation('common');
  const { t: errorT } = useTranslation('errors');
  const { setDirty } = useWorkspaceLayout();
  const { appConfig, onAppConfigUpdate } = useAppConfig(workspaceId, appId);

  const onSubmit = async (value: any) => {
    try {
      await onAppConfigUpdate(value);
      notification.success({
        message: t('apps.saveSuccess'),
        placement: 'bottomRight',
      });
      setDirty(false);
    } catch (e) {
      notification.error({
        message: errorT('unknown', { errorName: e }),
        placement: 'bottomRight',
      });
    }
  };

  const onChange = useCallback(
    (value: any) => {
      const keys = Object.keys(value);
      const prevValue = keys.reduce(
        (prev, key) => ({
          ...prev,
          [key]: appConfig[key],
        }),
        {}
      );
      if (JSON.stringify(prevValue) === JSON.stringify(value)) return;
      setDirty(true);
    },
    [appConfig, setDirty]
  );

  if (!appConfig) return <Loading />;

  if (schema) {
    const s: Schema = {
      type: 'object',
      properties: schema as Schema['properties'],
    };

    return (
      <div className="p-6">
        <SchemaForm
          schema={s}
          onChange={onChange}
          onSubmit={onSubmit}
          initialValues={appConfig}
          buttons={[
            <div className="flex w-full justify-end mt-5" key="submit">
              <Button
                type="submit"
                className="!p-1"
                variant="primary"
                key="submit"
              >
                {commmonT('save')}
              </Button>
            </div>,
          ]}
        />
      </div>
    );
  }
  if (block) {
    return (
      <BlockLoader
        api={api}
        url={block}
        token={`${api.token}`}
        workspaceId={workspaceId}
        appInstance={appId}
        config={{}}
        appConfig={appConfig}
        onAppConfigUpdate={onAppConfigUpdate}
      />
    );
  }

  return null;
};

export default AppEditor;
