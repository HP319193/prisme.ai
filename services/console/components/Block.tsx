import * as React from 'react';
import * as prismeaiDS from '@prisme.ai/design-system';
import * as prismeaiSDK from '../utils/api';
import { ReactElement, useEffect, useState } from 'react';
import { nanoid } from 'nanoid';
import { useTranslation } from 'next-i18next';

if (process.browser) {
  // @ts-ignore
  window.__external = window.__external || {
    React: { ...React, default: React },
    prismeaiDS,
    prismeaiSDK,
  };
}

class BlockErrorBoundary extends React.Component {
  state = {
    hasError: false,
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error: any, errorInfo: any) {
    // Vous pouvez aussi enregistrer l'erreur au sein d'un service de rapport.
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // Vous pouvez afficher n'importe quelle UI de repli.
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children;
  }
}

interface BlockComponentProps {
  entityId: string;
  token?: string;
  workspaceId?: string;
  appInstance?: string;
  language?: string;
}
type BlockComponent = (props: BlockComponentProps) => ReactElement;

interface BlockProps extends BlockComponentProps {
  url: string;
  renderLoading?: ReactElement;
}

export const ReactBlock = ({ url, renderLoading, ...props }: BlockProps) => {
  const {
    i18n: { language },
  } = useTranslation('workspaces');
  const [loading, setLoading] = useState(true);

  const [Component, setComponent] = useState<BlockComponent | null>(null);
  useEffect(() => {
    const uniqMethod = `__load_${nanoid()}`;
    // @ts-ignore
    window[uniqMethod] = (module) => {
      setComponent(() => {
        return module.default;
      });
      setLoading(false);
    };
    const s = document.createElement('script');

    s.innerText = `
    import * as module from '${url}';
    try {
      window['${uniqMethod}'](module);
    } catch (e) {}
    `;
    s.type = 'module';
    document.body.appendChild(s);

    return () => {
      // @ts-ignore
      delete window[uniqMethod];
    };
  }, [url]);

  return (
    <>
      {(loading && renderLoading) || null}
      {Component && <Component {...props} language={language} />}
    </>
  );
};

export const IFrameBlock = ({
  url,
  entityId,
  token,
  renderLoading,
}: BlockProps) => {
  const [loading, setLoading] = useState(true);
  const [height, setHeight] = useState(100);
  const handleLoad = React.useCallback(
    (e) => {
      setLoading(false);
      e.target.contentWindow.postMessage(
        {
          source: 'prisme.ai',
          token,
          id: entityId,
        },
        '*'
      );
    },
    [entityId, token]
  );
  useEffect(() => {
    const listener = (e: MessageEvent) => {
      if (e.data.source === entityId) {
        Object.keys(e.data).forEach((method) => {
          switch (method) {
            case 'init':
              setHeight(+e.data[method].height);
          }
        });
      }
    };
    window.addEventListener('message', listener);

    return () => {
      window.removeEventListener('message', listener);
    };
  }, [entityId, height]);

  return (
    <>
      {(loading && renderLoading) || null}
      <iframe
        src={url}
        onLoad={handleLoad}
        className="flex"
        style={{
          width: '100%',
          height: `${height}px`,
        }}
      />
    </>
  );
};

export const Block = ({ url, ...props }: BlockProps) => {
  const isJs = url.match(/\.js$/);
  if (isJs) {
    return (
      <BlockErrorBoundary>
        <ReactBlock url={url} {...props} />
      </BlockErrorBoundary>
    );
  }
  return <IFrameBlock url={url} {...props} />;
};

export default Block;
