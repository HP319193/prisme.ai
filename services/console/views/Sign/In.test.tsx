import SignIn from "./In";
import renderer, { act } from "react-test-renderer";
import { useUser } from "../../components/UserProvider";
import ApiError from "../../api/ApiError";
import { Form } from "react-final-form";
import { useRouter } from "next/router";

jest.mock("../../components/UserProvider", () => {
  const mock: any = {};
  mock.mock = mock;
  return {
    useUser: () => mock,
  };
});
jest.mock("next/router", () => {
  const push = jest.fn();
  return {
    useRouter: () => ({
      push,
    }),
  };
});

beforeEach(() => {
  const { mock } = useUser() as any;
  mock.user = null;
  mock.loading = false;
  mock.error = null;
  mock.signin = jest.fn();
});

it("should render", () => {
  const root = renderer.create(<SignIn />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should load", () => {
  (useUser() as any).mock.loading = true;
  const root = renderer.create(<SignIn />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should have error", () => {
  (useUser() as any).mock.error = new ApiError(
    {
      error: "Error",
      message: "Error",
    },
    400
  );
  const root = renderer.create(<SignIn />);
  expect(root.toJSON()).toMatchSnapshot();
});

it("should submit form", async () => {
  const root = renderer.create(<SignIn />);
  await act(async () => {
    await root.root
      .findByType(Form)
      .props.onSubmit({ username: "email@company.com", password: "123456" });
  });
  expect(useUser().signin).toHaveBeenCalledWith("email@company.com", "123456");
});

it("should validate form", async () => {
  const root = renderer.create(<SignIn />);
  expect(
    root.root.findByType(Form).props.validate({ username: "", password: "" })
  ).toEqual({
    username: "required",
    password: "required",
  });
  expect(
    root.root.findByType(Form).props.validate({ username: "a", password: "" })
  ).toEqual({
    password: "required",
  });
  expect(
    root.root.findByType(Form).props.validate({ username: "", password: "a" })
  ).toEqual({
    username: "required",
  });
  expect(
    root.root.findByType(Form).props.validate({ username: "a", password: "a" })
  ).toEqual({});
});

it("should redirect after signin", () => {
  const root = renderer.create(<SignIn />);
  act(() => {
    (useUser() as any).mock.user = {};
    root.update(<SignIn />);
  });
  expect(useRouter().push).toHaveBeenCalledWith("/workspaces");
});
