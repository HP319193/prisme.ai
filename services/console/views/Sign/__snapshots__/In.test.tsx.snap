// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have error 1`] = `
<div
  className="flex flex-1 flex-col-reverse md:flex-row overflow-y-auto"
>
  <div
    className="!flex flex-col bg-gradient-to-br from-[#0A1D3B] to-[#0F2A54] items-center justify-center md:w-[40vw]"
  >
    <div
      className="flex invisible md:visible flex-col flex-1 justify-center w-full space-y-4 lg:space-y-6"
    >
      <div
        className="w-1/2 h-4 lg:h-8 bg-accent rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-orange rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-grey rounded-r-[100rem]"
      />
    </div>
    <div
      className="flex flex-1 flex-col justify-end mb-10"
    >
      <div
        className="flex flex-1 items-center flex-col text-white"
      >
        <div
          className="flex flex-1 flex-col font-normal text-[1rem] md:text-[2rem] xl:text-[3.375rem] leading-normal"
        >
          <div
            className="flex flex-1 flex-col justify-center"
          />
          <div
            className="flex flex-row  mt-20"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="!flex flex-1 items-center justify-center md:w-[60vw] md:h-[100vh]"
  >
    <div
      className="flex flex-col items-center"
    >
      <div
        className="flex flex-col items-center space-y-4 mb-16 mt-8"
      >
        <div
          className="text-accent !font-light tracking-[.3em] uppercase text-xs"
        >
          in.topForm1
        </div>
        in.topForm2
        <div
          className="text-center"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should load 1`] = `
<div
  className="flex flex-1 flex-col-reverse md:flex-row overflow-y-auto"
>
  <div
    className="!flex flex-col bg-gradient-to-br from-[#0A1D3B] to-[#0F2A54] items-center justify-center md:w-[40vw]"
  >
    <div
      className="flex invisible md:visible flex-col flex-1 justify-center w-full space-y-4 lg:space-y-6"
    >
      <div
        className="w-1/2 h-4 lg:h-8 bg-accent rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-orange rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-grey rounded-r-[100rem]"
      />
    </div>
    <div
      className="flex flex-1 flex-col justify-end mb-10"
    >
      <div
        className="flex flex-1 items-center flex-col text-white"
      >
        <div
          className="flex flex-1 flex-col font-normal text-[1rem] md:text-[2rem] xl:text-[3.375rem] leading-normal"
        >
          <div
            className="flex flex-1 flex-col justify-center"
          />
          <div
            className="flex flex-row  mt-20"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="!flex flex-1 items-center justify-center md:w-[60vw] md:h-[100vh]"
  >
    <div
      className="flex flex-col items-center"
    >
      <div
        className="flex flex-col items-center space-y-4 mb-16 mt-8"
      >
        <div
          className="text-accent !font-light tracking-[.3em] uppercase text-xs"
        >
          in.topForm1
        </div>
        in.topForm2
        <div
          className="text-center"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render 1`] = `
<div
  className="flex flex-1 flex-col-reverse md:flex-row overflow-y-auto"
>
  <div
    className="!flex flex-col bg-gradient-to-br from-[#0A1D3B] to-[#0F2A54] items-center justify-center md:w-[40vw]"
  >
    <div
      className="flex invisible md:visible flex-col flex-1 justify-center w-full space-y-4 lg:space-y-6"
    >
      <div
        className="w-1/2 h-4 lg:h-8 bg-accent rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-orange rounded-r-[100rem]"
      />
      <div
        className="w-1/2 h-4 lg:h-8 bg-pr-grey rounded-r-[100rem]"
      />
    </div>
    <div
      className="flex flex-1 flex-col justify-end mb-10"
    >
      <div
        className="flex flex-1 items-center flex-col text-white"
      >
        <div
          className="flex flex-1 flex-col font-normal text-[1rem] md:text-[2rem] xl:text-[3.375rem] leading-normal"
        >
          <div
            className="flex flex-1 flex-col justify-center"
          />
          <div
            className="flex flex-row  mt-20"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="!flex flex-1 items-center justify-center md:w-[60vw] md:h-[100vh]"
  >
    <div
      className="flex flex-col items-center"
    >
      <div
        className="flex flex-col items-center space-y-4 mb-16 mt-8"
      >
        <div
          className="text-accent !font-light tracking-[.3em] uppercase text-xs"
        >
          in.topForm1
        </div>
        in.topForm2
        <div
          className="text-center"
        />
      </div>
    </div>
  </div>
</div>
`;
